[inductor] Make triton kernel autotune config defaults backward-compatible

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145494

If a model was torch.packaged using triton<=3.1, any user-defined
autotuned kernels will have reps/warmups burned in with the old defaults
(100/25).  If this model is loaded with triton>=3.2, inductor's checks for
unsupported non-default autotune args will fail, because triton.Autotuner's
defaults for these parameters has changed to `None`.  Let's explicitly support
those values for backward compatibility with these older models.

Differential Revision: [D68561014](https://our.internmc.facebook.com/intern/diff/D68561014/)

bertmaher committed:
[inductor] Make triton kernel autotune config defaults backward-compatible

If a model was torch.packaged using triton<=3.1, any user-defined
autotuned kernels will have reps/warmups burned in with the old defaults
(100/25).  If this model is loaded with triton>=3.2, inductor's checks for
unsupported non-default autotune args will fail, because triton.Autotuner's
defaults for these parameters has changed to `None`.  Let's explicitly support
those values for backward compatibility with these older models.

Differential Revision: [D68561014](https://our.internmc.facebook.com/intern/diff/D68561014/)

[ghstack-poisoned]

aorenste commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1144,25 +1144,21 @@ def init_variable(
             # Newer version of triton change attribute name from warmup to num_warmup and rep to num_rep.
             # The call to get_first_attr is to maintain backward-compatibility.
+
+            def defaults_ok(attr, alternates, values):
```
Any chance to include type annotations?

aorenste commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -1144,25 +1144,21 @@ def init_variable(
-                        "use_cuda_graph" in defaults
-                        and defaults["use_cuda_graph"].default != kernel.use_cuda_graph
-                    )
+                    not defaults_ok("num_warmups", ["warmup"], [25, None])
+                    or not defaults_ok("num_reps", ["rep"], [100, None])
+                    or not defaults_ok("use_cuda_graph", [], [False])
```
nit: As a general rule you should prefer tuples for constants, not list (because list causes an extra copy)
```suggestion
                    not defaults_ok("num_warmups", ("warmup",), (25, None))
                    or not defaults_ok("num_reps", ("rep",), (100, None))
                    or not defaults_ok("use_cuda_graph", (), (False,))
```

bertmaher committed:
Update on "[inductor] Make triton kernel autotune config defaults backward-compatible"


If a model was torch.packaged using triton<=3.1, any user-defined
autotuned kernels will have reps/warmups burned in with the old defaults
(100/25).  If this model is loaded with triton>=3.2, inductor's checks for
unsupported non-default autotune args will fail, because triton.Autotuner's
defaults for these parameters has changed to `None`.  Let's explicitly support
those values for backward compatibility with these older models.

Differential Revision: [D68561014](https://our.internmc.facebook.com/intern/diff/D68561014/)

[ghstack-poisoned]

bertmaher committed:
Update on "[inductor] Make triton kernel autotune config defaults backward-compatible"


If a model was torch.packaged using triton<=3.1, any user-defined
autotuned kernels will have reps/warmups burned in with the old defaults
(100/25).  If this model is loaded with triton>=3.2, inductor's checks for
unsupported non-default autotune args will fail, because triton.Autotuner's
defaults for these parameters has changed to `None`.  Let's explicitly support
those values for backward compatibility with these older models.

Differential Revision: [D68561014](https://our.internmc.facebook.com/intern/diff/D68561014/)

[ghstack-poisoned]

bertmaher commented:
Ugh, I keep messing up the typing and I can't figure out how to run mypy locally and this is going to turn into a SEV if I don't land it now, so going to land and fix up the typing in a follow-up

aorenste commented:
> Ugh, I keep messing up the typing and I can't figure out how to run mypy locally and this is going to turn into a SEV if I don't land it now, so going to land and fix up the typing in a follow-up

You should be able to run mypy locally with `lintrunner --all-files --take MYPY`

bertmaher committed:
Fix mypy, hopefully on "[inductor] Make triton kernel autotune config defaults backward-compatible"


If a model was torch.packaged using triton<=3.1, any user-defined
autotuned kernels will have reps/warmups burned in with the old defaults
(100/25).  If this model is loaded with triton>=3.2, inductor's checks for
unsupported non-default autotune args will fail, because triton.Autotuner's
defaults for these parameters has changed to `None`.  Let's explicitly support
those values for backward compatibility with these older models.

Differential Revision: [D68561014](https://our.internmc.facebook.com/intern/diff/D68561014/)

[ghstack-poisoned]

bertmaher commented:
@pytorchbot merge

bertmaher committed:
Update on "[inductor] Make triton kernel autotune config defaults backward-compatible"


If a model was torch.packaged using triton<=3.1, any user-defined
autotuned kernels will have reps/warmups burned in with the old defaults
(100/25).  If this model is loaded with triton>=3.2, inductor's checks for
unsupported non-default autotune args will fail, because triton.Autotuner's
defaults for these parameters has changed to `None`.  Let's explicitly support
those values for backward compatibility with these older models.

Differential Revision: [D68561014](https://our.internmc.facebook.com/intern/diff/D68561014/)

[ghstack-poisoned]

bertmaher commented:
@pytorchbot merge
